### ISSUE TITLE
AI Infrastructure for Endoregdb

### ISSUE THREAD

**Issue by Hamzaukw** [wg-lux/endoreg-db#302]
### TASK 1. MODEL OBJECT : AIDATASET

**Comment by Hamzaukw** [comment]
# Model AIDataSet

- AIDataSet is a Django model that describes a training dataset definition for AI models, not the raw tensors or vectors.
- It stores which AI model family the dataset belongs to via ai_model_type (for now defaulting to "image_multilabel_classification").
- It stores what kind of input data the dataset uses via dataset_type, currently only "image", with planned extensions for "video" and "text".
- For image datasets, it links to the concrete annotation rows through the image_annotations ManyToMany field, which points to ImageClassificationAnnotation objects, each connecting a frame and a label.
- The model is designed to later support separate ManyToMany relations for video and text annotations (commented placeholders for video_annotations and text_annotations), so each dataset_type uses its own link table.
- It has basic metadata fields name and description so you can give each dataset a human-readable identifier and documentation.
- It tracks lifecycle information with created_at and updated_at timestamps and can be soft-enabled or disabled for training using the is_active boolean.
- The helper method get_annotations_queryset() hides the underlying relation by returning the “active” annotation manager depending on dataset_type (currently image_annotations, later video/text), so the rest of the code can stay generic.
- The str method returns a readable string that includes the id and, if available, the name of the dataset.
- 

## Correct field list for AIDataSet


AIDataSet – corrected field list

### id

- PK.

### name

- Type: string (nullable).
- Purpose: human-readable identifier (e.g. "Legacy multilabel dataset v1").

### description

- Type: text (nullable).
- Purpose: optional notes / explanation.

### ai_model_type

- Type: string/choices.
- Examples: "image_multilabel_classification" (current model)
- Later: "video_multilabel", "text_classifier", etc.
- Purpose:
    -     Tells the system which AI model family this dataset is for.
    -     Used to pick correct architecture and output dimension logic.

### dataset_type

- Type: string/choices.
- Examples:"image" (current), later "video", "text", …
- Purpose:
    -     Controls which annotation table the ReqSet will use:

- Also tells your data loader what kind of input objects it will see.

### reqset field is missing

- Type: FK → imageclassificationannotation (iff datset_type is frames)
-     -     If "frames" → use ImageClassificationAnnotation (with frame_id, label_id).
    -     If "video" → future video-annotation table.
    -     If "text" → future text-annotation table.
- Purpose:
  -      Defines how to select annotations and labelsets from the DB.
- For dataset_type == "image":
  -      ReqSet selects a set of ImageClassificationAnnotation rows.
- Each row gives you:
  -      frame_id → which frame.
  -      label_id → which label.
- From these, you derive:
   -      The frames (and thus image paths).
   -      The labels, and thereby their LabelSet.


ADD THIS TO DOCUMENTATION FOR FUTURE USE

Earlier, the AIDataSet model used a single Many-To-Many field named reqset that always pointed to ImageClassificationAnnotation. This worked only for image datasets but made it impossible to add video or text annotations later because a Many-To-Many field cannot change which table it links to. The fix was to replace reqset with separate annotation fields: right now only image_annotations (M2M to ImageClassificationAnnotation), and later you will add video_annotations or text_annotations. To unify access, the model now includes a helper method get_annotations_queryset(), which automatically returns the correct annotation manager based on dataset_type.


For video and text we keep the same idea, just with separate fields and tables. For videos, the AIDataSet model will later get a video_annotations ManyToMany field that points to a video-annotation model (for example VideoSegmentationAnnotation). That creates its own join table in the database, separate from the image one. For text, we will add a text_annotations ManyToMany field pointing to a text-annotation model, again with its own join table. So in the database there will be three different link tables: one for image_annotations, one for video_annotations, one for text_annotations. On each AIDataSet row, only the field that matches dataset_type is actually used: if dataset_type="image" we fill image_annotations, if later dataset_type="video" we will fill video_annotations, and so on. The helper method get_annotations_queryset() hides this difference: it checks dataset_type and returns the right relation (image/video/text), so the dataloader and training code can always just call that helper without caring which concrete table is used underneath

**Comment by Hamzaukw** [comment]
## LOAD_LEGACY_DATA.PY

WHAT IT DOES

- Validates the paths for the JSONL file and images directory.
- Loads an existing VideoFile using the provided video ID.
- Loads an existing LabelSet by name and version.
- Opens the JSONL file line by line.
- Parses each JSON line into a Python dictionary.
- Extracts the list of labels for that image.
- Extracts the filename for the image.
- Constructs the absolute path to the image file.
- Warns if the image file does not physically exist.
- Creates a new Frame object linked to the chosen VideoFile.
- Assigns increasing frame numbers sequentially.
- Saves each Frame if not in dry-run mode.
- Loops through all labels listed for this frame.
- Ensures each label exists in the Label table (create if missing).
- Ensures each label is attached to the selected LabelSet.
- Creates an ImageClassificationAnnotation for each label.
- Sets annotation.value=True because legacy labels are all positives.
- Tags annotation.annotator="legacy_import".
- Saves each annotation unless dry-run is active.
- If AIDataSet is enabled (not dry-run), attaches annotation to ai_dataset.reqset.
- Repeats these steps for all lines in the JSONL file.
- Prints a final summary with total frames and annotation count, and indicates whether changes were committed or skipped via dry-run.

### Current comand to populate database objects

```bash
python manage.py load_legacy_data \                       
  --video-id 1 \
  --jsonl-path data/legacy_data/legacy_img_dicts.jsonl \
  --images-root data/legacy_data/images \
  --dataset-name legacy_multilabel_dataset_v1

```

**Comment by Hamzaukw** [comment]
## CHECK INPUT FOR DATALOADER

**Comment by Hamzaukw** [comment]
# QUESTION
---
## About 1 (true)

- In the table ```ImageClassificationAnnotation```, there is a field called ```value```.
- If for a given ```(frame_id, label_id)``` the value is ```True```, then I understand this means: **this label is present on this frame**
- so in the final vector it should be ```1```

### Example:

- Frame 10 has annotation row:

```python
frame_id = 10  
label_id = polyp  
value = True

```
Then output should be: ```polyp = 1 ```,  **#### Is this correct?**

---

## About 0 (false)

Now here I am confused. If ```value=False in ImageClassificationAnnotation```, for given ```(frame_id, label_id)``` does this mean:

- this label is explicitly not present for this frame
- so output should be ```0```?
- Example:-

```python
frame_id = 10  
label_id = outside  
value = False
```

- Then in final vector: ```coutside = 0```
- ```Is this what you mean by 0?```, Or do you mean something else?

---

##  About None (unknown)

Let’s say my LabelSet has 5 labels:

```bash
polyp, nbi, outside, ileum, blood
```
But the legacy input JSON for one frame only includes:

```python
["polyp", "nbi", "outside"]

```
And nothing about ```ileum``` and ```blood```.

write now, i treat like this:-

```bash
polyp : 1
nbi : 1
outside : 1
ileum : None (unknown)
blood : None (unknown)

```

Because there is no annotation row in the ```ImageClassificationAnnotation``` table for those missing labels.

```Is this the correct interpretation?```

**Comment by Hamzaukw** [comment]
[ResNet50 pretrained with DINOv1 on GastroNet-5M](https://huggingface.co/tgwboers/GastroNet-5M_Pretrained_Weights/tree/main)

**Comment by Hamzaukw** [comment]
### POINT OF CONCERN

Previously, for each frame and label we had three states:
```1 = present, 0 = explicitly not present, None = unknown / not annotated.```
In the first version of the training pipeline we treated ```None as “unknown”``` and set the ```mask to 0```, so the loss and metrics completely ignored those entries. Because our legacy JSON/data only ever explicitly writes ```positives (1) and never writes explicit negatives (0)```, this meant that after filtering to the ```v2 label set, almost every known entry was a positive```. The model never really saw proper negatives, so it could ```predict “positive” everywhere``` and still get artificially perfect metrics (precision/recall/F1 ≈ 1.0), which is ```misleading and useless``` for real evaluation.

Now, with the ```“unlabeled → negative”``` strategy for the ```v2 labels```, we do the following after filtering to those 9 labels:

If a label is in the JSON/DATABASE : we keep it as ```1 (positive)```.

If a label is not in the JSON for that frame : we treat it as ```0 (negative)``` and mark it as ```known ```in the mask.

This turns the v2 label space into a fully ```supervised 0/1 matrix```: the model sees both positives and negatives for each label, the loss can learn real decision boundaries, and the validation/test metrics become meaningful (you’ll see non-trivial precision/recall/F1 instead of fake 1.0). The trade-off is that we assume unlabeled = truly absent, which can introduce a bit of label noise if some positives were never annotated - but it’s a practical and common approach when we only have positive labels available.

NEED TO HANDLE if all positives in dataset and also situation when we have known 1s and 0s.

**Comment by Hamzaukw** [comment]
Almost completed - few code revisions required

```bash
 python manage.py model_input \                                                                                         
  --dataset-id 1 \
  --backbone-checkpoint data/model_training/checkpoints/RN50_GastroNet-1M_DINOv1.pth \
  --epochs 15
```

**Comment by Hamzaukw** [comment]
- We train a multi-label classifier on colonoscopy frames.
- The frames and labels live in the database, grouped by old_examination_id 
- We use an AIDataSet to select which frames and labels to use.
- For each frame, we build a 24-length label vector with 1, 0, or None base on the length of labels
- Then we filter to the v2 label set (9 labels) and apply the rule:
- for v2 labels, label is not annotated, we treat it as absent (0).
- We then split data into train/val/test by examination, so a single exam never appears in multiple splits.
- We wrap everything in a PyTorch Dataset that loads and normalizes images and returns (image, labels, mask).
- The model is a ResNet50 pretrained on GastroNet, where we remove the final layer and add a new 9-output classifier.
- We freeze the backbone and train only the new head using AdamW.
- Loss is multi-label focal loss with class weights and masks, which helps with unbalanced labels and hard examples.
- We use a warm-up + cosine learning rate schedule to gradually increase and then decrease learning rate, which stabilizes training.
- During and after training, we compute global precision/recall/F1 and also per-label precision/recall/F1/support and print them in tables.
- This shows that labels like ‘outside’ and ‘low_quality’ perform very well, while ‘polyp’ has good precision but low recall, meaning we miss many real polyps and need to work further on that label (e.g. data, thresholding, or loss tuning


#bit changes had been made, need to update it for final documentation

**Comment by Hamzaukw** [comment]
python manage.py model_input \
  --dataset-id 1 \
  --epochs 10 \
  --backbone-name resnet50_imagenet
Note: no --backbone-checkpoint



python manage.py model_input \
  --dataset-id 1 \
  --backbone-checkpoint data/model_training/checkpoints/RN50_GastroNet-1M_DINOv1.pth \
  --epochs 5 \
  --backbone-name gastro_rn50

**Comment by Hamzaukw** [comment]
to add new model in the current architecture
endoreg_db/utils/ai/model_training/model_backbones.py
add 
```python
elif config.backbone_name == "efficientnet_b4":
    print("[Backbone] Using EfficientNet-B4 pretrained on ImageNet")

    weights = EfficientNet_B4_Weights.IMAGENET1K_V1
    model = efficientnet_b4(weights=weights)

    # Remove the classifier — EfficientNet’s last layer:
    in_features = model.classifier[1].in_features

    # Freeze backbone if you want (same as ResNet logic)
    for p in model.parameters():
        p.requires_grad = config.freeze_backbone

    backbone = nn.Sequential(
        model.features,
        model.avgpool,   # outputs [batch, 1792]
        nn.Flatten(1)
    )

    return backbone, in_features


```
endoreg_db/utils/ai/model_training/config.py
TrainingConfig

**Comment by Hamzaukw** [comment]
## Adding new model

Backbone name :  efficientnet_b4_imagenet

### Add the new backbone builder in model_backbones.py

endoreg_db/utils/ai/model_training/model_backbones.py

Right below _build_efficientnet_b0_backbone, add a B4 version 

```python
def _build_efficientnet_b4_backbone() -> Tuple[nn.Module, int]:
    """
    EfficientNet-B4 backbone with ImageNet weights.

    Returns:
      - backbone: feature extractor (no classifier)
      - in_features: feature dimension for the final linear head
    """
    from torchvision.models import efficientnet_b4, EfficientNet_B4_Weights

    # Load EfficientNet-B4 pretrained on ImageNet
    base = efficientnet_b4(weights=EfficientNet_B4_Weights.IMAGENET1K_V1)

    # base.features: [B, C, H, W]
    # We add a global average pool to get [B, C, 1, 1]
    features = base.features
    backbone = nn.Sequential(
        features,
        nn.AdaptiveAvgPool2d(1),  # -> [B, C, 1, 1]
    )

    # The classifier is usually: Sequential(Dropout, Linear(in_features, 1000))
    in_features = base.classifier[1].in_features

    return backbone, in_features

```

In the same file add into create_multilabel_model

```
    elif backbone_name == "efficientnet_b0_imagenet":
        backbone, in_features = _build_efficientnet_b0_backbone()
    elif backbone_name == "efficientnet_b4_imagenet":
        backbone, in_features = _build_efficientnet_b4_backbone()

```

(Optional ) Document it in TrainingConfig, In the TrainingConfig docstring near backbone_name
endoreg_db/utils/ai/model_training/config.py

Expose it in the management command help (endoreg_db/management/commands/model_input.py)
Find the add_arguments part for --backbone-name:

```python
        parser.add_argument(
            "--backbone-name",
            type=str,
            default="gastro_rn50",
            help=(
                "Backbone name, e.g. 'gastro_rn50' (default), "
               "'resnet50_imagenet', 'resnet50_random', 'efficientnet_b0_imagenet', etc."
        ),
    )

Update the help string to also mention B4:
            help=(
                "Backbone name, e.g. 'gastro_rn50' (default), "
                "'resnet50_imagenet', 'resnet50_random', "
                "'efficientnet_b0_imagenet', 'efficientnet_b4_imagenet', etc."
            ),



```

Think about image size for B4 (b4 native size is 380) so check for each model

```bash
python manage.py model_input \
    --dataset-id 1 \
    --backbone-name efficientnet_b4_imagenet \
    --epochs 10


```

Do NOT pass --backbone-checkpoint for EfficientNet-B4 — it uses ImageNet weights from torchvision, not GastroNet.

The checkpoint argument is only meaningful for "gastro_rn50".

model will be saved as something like:data/model_training/runs/aidataset_1_efficientnet_b4_imagenet_v2_multilabel.pth

**Comment by Hamzaukw** [comment]
# Global Metrics Comparison

## Comparison of ResNet50 on ImageNet Data and ResNet50 on GastroNet5M Data

### Final Test Metrics for 5 Epochs

- ```Precision``` = when the model says ```this label is present```, how often is it correct?
- ```Recall``` = how many of the real positives it actually finds.
- ```F1``` = a balance between precision and recall. the model is both correct when it predicts (precision) and it finds many of the true cases (recall).
- ```Accuracy``` = how often it’s correct overall (includes “no label” cases).

| Model                     | Precision | Recall | F1     | Accuracy |
|---------------------------|-----------|--------|--------|----------|
| ResNet50 ImageNet         | 0.8717    | 0.3869 | 0.5359 | 0.9239   |
| GastroNet RN50 checkpoint | 0.8642    | 0.5482 | 0.6708 | 0.9389   |


### Per-label comparison : final validation per-label F1 (epoch 5)

- ```Final Validation per-label F1``` = how good the model is on the validation set, using the F1 score for that specific label.

| Label           | ResNet50 ImageNet F1 | GastroNet F1 |
|-----------------|----------------------|--------------|
| appendix        | 0.27                 | **0.67**     |
| blood           | 0.45                 | **0.59**     |
| ileocaecalvalve | 0.40                 | **0.60**     |
| ileum           | 0.60                 | **0.80**     |
| low_quality     | 0.80                 | **0.88**     |
| outside         | 0.89                 | **0.96**     |
| **polyp**       | **0.14**             | **0.37**     |
| water_jet       | 0.47                 | **0.64**     |
| wound           | 0.19                 | **0.52**     |


#### Observation : ResNet50 on GastroNet-5M is better across the board.

@Maddonix @maxhild  @PeterKczyk

**Comment by Hamzaukw** [comment]
need to look for null entries in field for unknowns in  imageclassificationannotation table

**Comment by Hamzaukw** [comment]
Manual annotation of 20 videos for the following labels: clip, snare, grasper, needle, digital chromo, outside, waterjet, skin and ileum.
Low-quality labeling will be applied later on selected frames after the initial manual annotation (but precision value of low-quality should be included to make any rule later).

**Comment by Hamzaukw** [comment]
closing this issue as model has been developed and initially test and performed well on legacy dataset, secondly new repo Lx-ai is created for ai model.
